mathjax3-ts/input/tex/FilterUtil.ts

```diff
@@ -0,0 +1,174 @@
+ */
+
+
+import {TEXCLASS, MmlNode, TextNode} from '../../core/MmlTree/MmlNode.js';
+import NodeUtil from './NodeUtil.js';
+import ParseOptions from './ParseOptions.js';
+import {MmlMo} from '../../core/MmlTree/MmlNodes/mo.js';
+
+
+namespace FilterUtil {
+
+  /**
+   * Visitor to set stretchy attributes to false on <mo> elements, if they are
+   * not used as delimiters. Also wraps non-stretchy infix delimiters into a
+   * TeXAtom.
+   * @param {MmlNode} node The node to rewrite.
+   * @param {ParseOptions} options The parse options.
+   */
+  export let cleanStretchy = function(arg: {math: any, data: ParseOptions}) {
+    let options = arg.data;
+    for (let mo of options.getList('fixStretchy')) {
+      if (NodeUtil.getProperty(mo, 'fixStretchy')) {
+        let symbol = NodeUtil.getForm(mo);
+        if (symbol && symbol[3] && symbol[3]['stretchy']) {
+          NodeUtil.setAttribute(mo, 'stretchy', false);
+        }
+        const parent = mo.parent;
+        if (!NodeUtil.getTexClass(mo) && (!symbol || !symbol[2])) {
+          const texAtom = options.nodeFactory.create('node', 'TeXAtom', [mo]);
+          parent.replaceChild(texAtom, mo);
+          texAtom.setInheritedAttributes({}, arg.math['display'],
+                                         parent.attributes.get('scriptlevel') as number,
+                                         parent.getProperty('texprimestyle') as boolean);
+        }
+        NodeUtil.removeProperties(mo, 'fixStretchy');
+      }
+    }
+  };
+
+
+  /**
+   * Visitor that removes superfluous attributes from nodes. I.e., if a node has
+   * an attribute, which is also an inherited attribute it will be removed. This
+   * is necessary as attributes are set bottom up in the parser.
+   * @param {MmlNode} mml The node to clean.
+   * @param {ParseOptions} options The parse options.
+   */
+  // TODO (DC): Move this maybe into setInheritedAttributes method?
+  export let cleanAttributes = function(arg: {data: ParseOptions}) {
+    let node = arg.data.root as MmlNode;
+    node.walkTree((mml: MmlNode, d: any) => {
+      let attribs = mml.attributes as any;
+      for (const key of attribs.getExplicitNames()) {
+        if (attribs.attributes[key] === mml.attributes.getInherited(key)) {
+          delete attribs.attributes[key];
+        }
+      }
+    }, {});
+  };
+
+
+  /**
+   * Combine adjacent <mo> elements that are relations (since MathML treats the
+   * spacing very differently)
+   * @param {MmlNode} mml The node in which to combine relations.
+   * @param {ParseOptions} options The parse options.
+   */
+  export let combineRelations = function(arg: {data: ParseOptions}) {
+    for (let mo of arg.data.getList('mo')) {
+      if (mo.getProperty('relationsCombined') || !mo.parent ||
+          (mo.parent && !NodeUtil.isType(mo.parent, 'mrow'))) {
+        continue;
+      }
+      let mml = mo.parent;
+      let m1: MmlNode, m2: MmlNode;
+      let children = mml.childNodes as (MmlNode|TextNode)[];
```

Comment:
There's a couple of problems with this:
* I can't just take the next element in the `mo` node list as it might the next sibling might be anywhere in the list (e.g., if it has been rewritten by some fluke later on).
* `childPosition()` won't work either as --- from my understanding --- it skips `inferredMrow` elements. So in most cases I get a `0` for `childPosition`.
* I also figured that using `childPosition` might not be a good idea if I have to call it multiple times as it would go over the list of children every time, rather than going through the list once.

But I can see that I could only start at the actual position of the node in the parent's children, which should work with an `indexOf` and then check the next element(s) only.

Reply:
Sorry, I wasn't clear in my initial statement.  I didn't mean take the next mo in the mo list, but the next element in the `childNodes` of the mo's parent.  I forgot that `childPosition()` skips inferred rows.  Maybe we need a version that doesn't do that (like your `indexOf` approach).